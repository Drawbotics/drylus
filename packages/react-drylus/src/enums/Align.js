import Enum from '@drawbotics/enums';


const Align = new Enum(
  'LEFT',
  'CENTER',
  'RIGHT',
);


export default Align;
