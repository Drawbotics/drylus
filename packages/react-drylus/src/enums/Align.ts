export enum Align {
  LEFT = 'LEFT',
  CENTER = 'CENTER',
  RIGHT = 'RIGHT',
  JUSTIFY = 'JUSTIFY',
}
