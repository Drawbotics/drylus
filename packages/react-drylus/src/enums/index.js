export Category from './Category';
export Tier from './Tier';
export Size from './Size';
export Position from './Position';
export Align from './Align';
export Color from './Color';
export Shade from './Shade';
