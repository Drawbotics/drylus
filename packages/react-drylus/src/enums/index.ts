export * from './Category';
export * from './Tier';
export * from './Size';
export * from './Position';
export * from './Align';
export * from './Color';
export * from './Shade';
