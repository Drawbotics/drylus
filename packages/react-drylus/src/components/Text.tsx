/* eslint-disable no-use-before-define */
import sv from '@drawbotics/drylus-style-vars';
import { css, cx } from 'emotion';
import isArray from 'lodash/isArray';
import isObject from 'lodash/isObject';
import React, { Fragment } from 'react';

import { Category, Color, Shade, Size, Tier } from '../enums';
import { Responsive, Style } from '../types';
import {
  ShowDateTime,
  colorEnumToCategory,
  generateDisplayedDate,
  generateDisplayedPrice,
  getEnumAsClass,
  shadeEnumToTier,
  useResponsiveProps,
} from '../utils';
import { TextLink } from './TextLink';

export {
  generateDisplayedPrice as formatPrice,
  ShowDateTime,
  generateDisplayedDate as formatDate,
} from '../utils';

const styles = {
  root: css`
    display: inline;
  `,
  bold: css`
    font-weight: 500;
    letter-spacing: 0.01rem;
  `,
  light: css`
    font-weight: 300;
    letter-spacing: 0.06rem;
  `,
  primary: css`
    color: ${sv.colorPrimary};
  `,
  secondary: css`
    color: ${sv.colorSecondary};
  `,
  tertiary: css`
    color: ${sv.colorTertiary};
  `,
  disabled: css`
    color: ${sv.colorDisabled};
  `,
  primaryInversed: css`
    color: ${sv.colorPrimaryInverse};
  `,
  secondaryInversed: css`
    color: ${sv.colorSecondaryInverse};
  `,
  tertiaryInversed: css`
    color: ${sv.colorTertiaryInverse};
  `,
  disabledInversed: css`
    color: ${sv.colorDisabledInverse};
  `,
  small: css`
    font-size: 0.9rem;
  `,
  default: css`
    font-size: 1rem;
  `,
  large: css`
    font-size: 1.1rem;
  `,
  brand: css`
    color: ${sv.brand};
  `,
  success: css`
    color: ${sv.success};
  `,
  danger: css`
    color: ${sv.danger};
  `,
  warning: css`
    color: ${sv.warning};
  `,
  info: css`
    color: ${sv.info};
  `,
};

type Price = {
  currency?: string;
  value: number;
};

export type TextChildren =
  | string
  | React.ReactElement<typeof Text>
  | React.ReactElement<typeof TextLink>
  | number
  | Price
  | Date
  | React.ReactNode;

export interface TextProps {
  /**
   * Makes the text visible on dark backgrounds
   * @default false
   */
  inversed?: boolean;

  /** @default false */
  bold?: boolean;

  /**
   * The opposite of bold, will set the font weight to 300 (useful with `inversed` on dark backgrounds)
   * @default false
   */
  light?: boolean;

  /**
   * @default Size.DEFAULT
   * @kind Size
   */
  size?: Size.DEFAULT | Size.SMALL | Size.LARGE;

  tier?: Tier;

  /** @default Shade.DARK */
  shade?: Shade;

  /**
   * Makes the text appear disabled, but still selectable
   * @default false
   */
  disabled?: boolean;

  children: TextChildren | Array<TextChildren> | React.ReactNode;

  /** @kind Category */
  category?: Category.BRAND | Category.SUCCESS | Category.INFO | Category.WARNING | Category.DANGER;

  /** @kind Color */
  color?: Color.BRAND | Color.RED | Color.BLUE | Color.GREEN | Color.ORANGE | Color.PRIMARY;

  /** Options to change the way the date is displayed, if provided. showTime toggles display of hour/minutes, format for dayjs overrides */
  dateOptions?: {
    showTime?: ShowDateTime;
    asArchive?: boolean;
    format?: any;
  };

  /** Formatting options for the .toLocaleString method used internally when formatting numbers */
  priceOptions?: any;

  /**
   * Used to override the current locale if necessary (e.g. if the browser locale is not explicitely defined)
   */
  locale?: string;

  /** Used for style overrides */
  style?: Style;

  /** Used for style overrides */
  className?: string;

  /** Reponsive prop overrides */
  responsive?: Responsive<this>;
}

function _processChild(
  child: TextChildren,
  {
    dateOptions,
    locale,
    priceOptions,
  }: {
    dateOptions: TextProps['dateOptions'];
    locale: TextProps['locale'];
    priceOptions: TextProps['priceOptions'];
  },
) {
  if (isObject(child)) {
    if (child instanceof Date) {
      return generateDisplayedDate({
        date: child,
        options: dateOptions,
        locale,
      });
    } else if ((child as Price).value != null) {
      return generateDisplayedPrice({
        price: child as Price,
        options: priceOptions,
        locale,
      });
    } else {
      return child;
    }
  }
  return child;
}

export const Text = ({ responsive, ...rest }: TextProps) => {
  const {
    inversed = false,
    bold = false,
    size = Size.DEFAULT,
    tier: _tier,
    disabled = false,
    children,
    category: _category,
    color,
    shade = Shade.DARK,
    style,
    light = false,
    dateOptions,
    priceOptions,
    locale,
    className,
  } = useResponsiveProps<TextProps>(rest, responsive);

  const tier = _tier ?? (shade ? shadeEnumToTier(shade) : null);
  const category = color ? colorEnumToCategory(color) : _category;

  const transformedChildren = isArray(children)
    ? [...children].map((child) =>
        _processChild(child, {
          dateOptions,
          locale,
          priceOptions,
        }),
      )
    : _processChild(children, { dateOptions, locale, priceOptions });

  return (
    <span
      className={cx(
        styles.root,
        {
          [styles.bold]: bold,
          [styles.light]: light,
          [styles.primary]: tier === Tier.PRIMARY && !disabled && !inversed,
          [styles.secondary]: tier === Tier.SECONDARY && !disabled && !inversed,
          [styles.tertiary]: tier === Tier.TERTIARY && !disabled && !inversed,
          [styles.disabled]: disabled && !inversed,
          [styles.primaryInversed]: tier === Tier.PRIMARY && !disabled && inversed,
          [styles.secondaryInversed]: tier === Tier.SECONDARY && !disabled && inversed,
          [styles.tertiaryInversed]: tier === Tier.TERTIARY && !disabled && inversed,
          [styles.disabledInversed]: disabled && inversed,
          [styles.small]: size === Size.SMALL,
          [styles.default]: size === Size.DEFAULT,
          [styles.large]: size === Size.LARGE,
          [styles[getEnumAsClass<typeof styles>(category)]]:
            category != null && !disabled && !inversed,
        },
        className,
      )}
      style={style}>
      <Fragment>{transformedChildren}</Fragment>
    </span>
  );
};
