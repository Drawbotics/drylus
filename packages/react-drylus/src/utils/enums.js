import camelCase from 'lodash/camelCase';

import { Category, Color, Shade, Tier } from '../enums';

export function getEnumAsClass(enumVal) {
  return camelCase(enumVal?.description?.toLowerCase());
}

export function colorEnumToCategory(enumVal) {
  switch (enumVal) {
    case Color.RED:
      return Category.DANGER;
    case Color.GREEN:
      return Category.SUCCESS;
    case Color.ORANGE:
      return Category.WARNING;
    case Color.BLUE:
      return Category.INFO;
  }
  return enumVal;
}

export function categoryEnumToColor(enumVal) {
  switch (enumVal) {
    case Category.DANGER:
      return Color.RED;
    case Category.SUCCESS:
      return Color.GREEN;
    case Category.WARNING:
      return Color.ORANGE;
    case Category.INFO:
      return Color.BLUE;
  }
  return enumVal;
}

export function shadeEnumToTier(enumVal) {
  switch (enumVal) {
    case Shade.DARK:
      return Tier.PRIMARY;
    case Shade.MEDIUM:
      return Tier.SECONDARY;
    case Shade.LIGHT:
      return Tier.TERTIARY;
  }
  return enumVal;
}

export function getIconForCategory(category) {
  switch (category) {
    case Category.DANGER:
      return 'alert-circle';
    case Category.SUCCESS:
      return 'check-circle';
    case Category.WARNING:
      return 'alert-triangle';
    default:
      return 'info';
  }
}
