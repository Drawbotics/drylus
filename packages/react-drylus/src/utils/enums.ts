import camelCase from 'lodash/camelCase';

import { Category, Color, Position, Shade, Size, Tier } from '../enums';

type Enum = Category | Color | Shade | Tier | Size | Position;

export function getEnumAsClass<T extends Record<string, string>>(enumVal?: Enum): keyof T {
  return camelCase(enumVal?.toLowerCase());
}

export function colorEnumToCategory(enumVal: Color): Category | undefined {
  switch (enumVal) {
    case Color.RED:
      return Category.DANGER;
    case Color.GREEN:
      return Category.SUCCESS;
    case Color.ORANGE:
      return Category.WARNING;
    case Color.BLUE:
      return Category.INFO;
  }
}

export function categoryEnumToColor(enumVal: Category): Color | undefined {
  switch (enumVal) {
    case Category.DANGER:
      return Color.RED;
    case Category.SUCCESS:
      return Color.GREEN;
    case Category.WARNING:
      return Color.ORANGE;
    case Category.INFO:
      return Color.BLUE;
  }
}

export function shadeEnumToTier(enumVal: Shade): Tier | undefined {
  switch (enumVal) {
    case Shade.DARK:
      return Tier.PRIMARY;
    case Shade.MEDIUM:
      return Tier.SECONDARY;
    case Shade.LIGHT:
      return Tier.TERTIARY;
  }
}

export function getIconForCategory(category: Category): string {
  switch (category) {
    case Category.DANGER:
      return 'alert-circle';
    case Category.SUCCESS:
      return 'check-circle';
    case Category.WARNING:
      return 'alert-triangle';
    default:
      return 'info';
  }
}
