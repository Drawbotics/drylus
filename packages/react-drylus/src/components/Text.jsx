import React from 'react';
import PropTypes from 'prop-types';
import { css, cx } from 'emotion';
import sv from '@drawbotics/drylus-style-vars';
import isObject from 'lodash/isObject';
import isArray from 'lodash/isArray';

import { Tier, Size, Category } from '../enums';
import { getEnumAsClass } from '../utils';
import { useResponsiveProps } from '../utils/hooks';
import { generateDisplayedDate, ShowDateTime } from '../utils/date/date';


const styles = {
  root: css`
    display: inline;
  `,
  bold: css`
    font-weight: 500;
    letter-spacing: 0.01rem;
  `,
  light: css`
    font-weight: 300;
    letter-spacing: 0.06rem;
  `,
  primary: css`
    color: ${sv.colorPrimary};
  `,
  secondary: css`
    color: ${sv.colorSecondary};
  `,
  tertiary: css`
    color: ${sv.colorTertiary};
  `,
  disabled: css`
    color: ${sv.colorDisabled};
  `,
  primaryInversed: css`
  color: ${sv.colorPrimaryInverse};
  `,
  secondaryInversed: css`
    color: ${sv.colorSecondaryInverse};
  `,
  tertiaryInversed: css`
    color: ${sv.colorTertiaryInverse};
  `,
  disabledInversed: css`
    color: ${sv.colorDisabledInverse};
  `,
  small: css`
    font-size: 0.9rem;
  `,
  default: css`
    font-size: 1rem;
  `,
  large: css`
    font-size: 1.1rem;
  `,
  brand: css`
    color: ${sv.brand};
  `,
  success: css`
    color: ${sv.success};
  `,
  danger: css`
    color: ${sv.danger};
  `,
  warning: css`
    color: ${sv.warning};
  `,
  info: css`
    color: ${sv.info};
  `,
};


function _processChild(child, dateOptions) {
  if (isObject(child)) {
    if (child instanceof Date) {
      // Handle dates
      return generateDisplayedDate({
        date: child,
        options: dateOptions,
      });
    }
    else if (child.value != null && child.currency != null) {
      // Handle currency
      return '[currency]';
    }
    else {
      console.warn('Unsupported Text child type. Please provde text, Date or Currency');
      return '';
    }
  }
  return child;
}


const Text = ({
  responsive,
  ...rest
}) => {
  const {
    inversed,
    bold,
    size,
    tier,
    disabled,
    children,
    category,
    style,
    light,
    dateOptions,
  } = useResponsiveProps(rest, responsive);

  const transformedChildren = isArray(children)
    ? [...children].map((child) => _processChild(child, dateOptions)).join('')
    : _processChild(children, dateOptions);

  return (
    <span className={cx(styles.root, {
      [styles.bold]: bold,
      [styles.light]: light,
      [styles.primary]: tier === Tier.PRIMARY && ! disabled && ! inversed,
      [styles.secondary]: tier === Tier.SECONDARY && ! disabled && ! inversed,
      [styles.tertiary]: tier === Tier.TERTIARY && ! disabled && ! inversed,
      [styles.disabled]: disabled && ! inversed,
      [styles.primaryInversed]: tier === Tier.PRIMARY && ! disabled && inversed,
      [styles.secondaryInversed]: tier === Tier.SECONDARY && ! disabled && inversed,
      [styles.tertiaryInversed]: tier === Tier.TERTIARY && ! disabled && inversed,
      [styles.disabledInversed]: disabled && inversed,
      [styles.small]: size === Size.SMALL,
      [styles.default]: size === Size.DEFAULT,
      [styles.large]: size === Size.LARGE,
      [styles[getEnumAsClass(category)]]: category && ! disabled && ! inversed,
    })} style={style}>
      {transformedChildren}
    </span>
  );
};


Text.propTypes = {
  /** Makes the text visible on dark backgrounds */
  inversed: PropTypes.bool,

  bold: PropTypes.bool,

  /** The opposite of bold, will set the font weight to 300 (useful with `inversed` on dark backgrounds) */
  light: PropTypes.bool,

  size: PropTypes.oneOf([Size.SMALL, Size.DEFAULT, Size.LARGE]),

  tier: PropTypes.oneOf([Tier.PRIMARY, Tier.SECONDARY, Tier.TERTIARY]),

  /** Makes the text appear disabled, but still selectable */
  disabled: PropTypes.bool,

  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.shape({
      currency: PropTypes.string.isRequired,
      value: PropTypes.number.isRequired,
    }),
    PropTypes.instanceOf(Date),
    PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
        PropTypes.shape({
          currency: PropTypes.string.isRequired,
          value: PropTypes.number.isRequired,
        }),
        PropTypes.instanceOf(Date),
      ])
    ),
  ]).isRequired,

  category: PropTypes.oneOf([
    Category.BRAND,
    Category.DANGER,
    Category.SUCCESS,
    Category.INFO,
    Category.WARNING,
  ]),

  /** Custom style object override */
  style: PropTypes.object,

  /** Reponsive prop overrides */
  responsive: PropTypes.shape({
    XS: PropTypes.object,
    S: PropTypes.object,
    M: PropTypes.object,
    L: PropTypes.object,
    XL: PropTypes.object,
    HUGE: PropTypes.object,
  }),
  
  /** Options to change the way the date is displayed, if provided. showTime toggles display of hour/minutes, format for dayjs overrides */
  dateOptions: PropTypes.shape({
    showTime: PropTypes.oneOf([
      ShowDateTime.DEFAULT,
      ShowDateTime.NEVER,
      ShowDateTime.ALWAYS,
    ]),
    asArchive: PropTypes.bool,
    // need to add options for dayjs
  }),
};


Text.defaultProps = {
  inversed: false,
  bold: false,
  light: false,
  size: Size.DEFAULT,
  tier: Tier.PRIMARY,
  disabled: false,
};


export default Text;
