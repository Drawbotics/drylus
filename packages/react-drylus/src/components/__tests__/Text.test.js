import React from 'react';
import { create } from 'react-test-renderer';
import get from 'lodash/get';

import { Category, Size } from '../../enums';
import Text from '../Text';


// Mock constant date is 01/06/2020 at 12pm
describe('Text', () => {
  describe('matches snapshot when it', () => {
    it('has children', () => {
      const tree = create(
        <Text>Text content</Text>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
    
    it('is bold', () => {
      const tree = create(
        <Text bold>Text content</Text>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('is light and inversed', () => {
      const tree = create(
        <Text light inversed>Text content</Text>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('has a size and category', () => {
      const tree = create(
        <Text size={Size.LARGE} category={Category.BRAND}>Text content</Text>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('matches expected output when it', () => {
    it('contains a date', () => {
      const date = new Date('2020-06-04 18:00');
  
      const tree = create(
        <Text>{date}</Text>
      ).toJSON();
      const children = get(tree, 'children[0]', '');
      
      expect(children).toEqual('Thu 4 Jun, 6:00 PM');
    });

    it('contains a price', () => {
     const price = { value: 10000, currency: 'GBP' };
  
      const tree = create(
        <Text>{price}</Text>
      ).toJSON();
      const children = get(tree, 'children[0]', '');
      
      expect(children).toEqual('Â£10,000');
    });
  });
});