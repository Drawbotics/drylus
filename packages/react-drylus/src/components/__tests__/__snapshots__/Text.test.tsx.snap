// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text matches snapshot when it has a size and category 1`] = `
.emotion-0 {
  display: inline;
  color: #172b4e;
  font-size: 1.1rem;
  color: #3DB5D0;
}

<span
  className="emotion-0"
>
  Text content
</span>
`;

exports[`Text matches snapshot when it has children 1`] = `
.emotion-0 {
  display: inline;
  color: #172b4e;
  font-size: 1rem;
}

<span
  className="emotion-0"
>
  Text content
</span>
`;

exports[`Text matches snapshot when it is bold 1`] = `
.emotion-0 {
  display: inline;
  font-weight: 500;
  -webkit-letter-spacing: 0.01rem;
  -moz-letter-spacing: 0.01rem;
  -ms-letter-spacing: 0.01rem;
  letter-spacing: 0.01rem;
  color: #172b4e;
  font-size: 1rem;
}

<span
  className="emotion-0"
>
  Text content
</span>
`;

exports[`Text matches snapshot when it is light and inversed 1`] = `
.emotion-0 {
  display: inline;
  font-weight: 300;
  -webkit-letter-spacing: 0.06rem;
  -moz-letter-spacing: 0.06rem;
  -ms-letter-spacing: 0.06rem;
  letter-spacing: 0.06rem;
  color: #fff;
  font-size: 1rem;
}

<span
  className="emotion-0"
>
  Text content
</span>
`;
